### PR TITLE
Too many Tocks in typed TimerSpec

### LINKED ISSUE #24360: Failed: TimerSpec
https://jenkins.akka.io:8498/job/pr-validator-per-commit-jenkins/11371/testReport/junit/

### PR DESCRIPTION
**Pull request by johanandren**
Fixes #24360

Adds a fishForMessages for typed teskit probes

### CODE REVIEW

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/TimerSpec.scala [hunk 1]
```diff
@@ -141,13 +144,20 @@ class TimerSpec extends TestKit("TimerSpec")
       val ref = spawn(behv)
       probe.expectMessage(Tock(1))
       ref ! Cancel
+      probe.fishForMessage(3.seconds) {
```
Comment (ktoso) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/TimerSpec.scala]: Shall we make it have a hint always? (string to explain what we wait for), I found that very helpful

File: akka-testkit-typed/src/main/scala/akka/testkit/typed/TestKit.scala [hunk 2]
```diff
@@ -17,6 +17,13 @@ import scala.util.control.NoStackTrace
  */
 case class TE(message: String) extends RuntimeException(message) with NoStackTrace
 
+/**
+ * Not for user extension.
+ *
+ * Instances are available from `FishingOutcomes` in the respective dsls
+ */
+@DoNotInherit abstract class FishingOutcome private[akka] ()
```
Comment (ktoso) [inline akka-testkit-typed/src/main/scala/akka/testkit/typed/TestKit.scala]: nice name.

"The Biggest Catch!" ;-)

Comment (ktoso) [inline akka-testkit-typed/src/main/scala/akka/testkit/typed/TestKit.scala]: link to scaladsl / javadsl so people know where from to obtain the impls?

File: akka-testkit-typed/src/main/scala/akka/testkit/typed/scaladsl/TestProbe.scala [hunk 3]
```diff
@@ -247,6 +266,52 @@ class TestProbe[M](name: String)(implicit system: ActorSystem[_]) {
...
+   * Additionally failures includes the list of messages consumed. If a message of type `M` but not of type `T` is
+   * received this will also fail the test, additionally if the `fisher` function throws a match error the error
+   * is decorated with some fishing details and the test is failed (making it convenient to use this method with a
+   * partial function).
+   *
+   * @param max Max total time without the fisher function returning `CompleteFishing` before failing
+   *            The timeout is dilated.
+   * @return The messages accepted in the order they arrived
+   */
+  def fishForMessage(max: FiniteDuration)(fisher: M ⇒ FishingOutcome): List[M] = {
```
Comment (ktoso) [inline akka-testkit-typed/src/main/scala/akka/testkit/typed/scaladsl/TestProbe.scala]: I'd like a `hint: String` I think, otherwise nice

**Review by ktoso — APPROVED** [review]
Minor request

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/TimerSpec.scala (hunk 1 above)
Comment (johanandren) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/TimerSpec.scala]: Hmm, you can put that in your failure message with FishingOutcomes.Fail, so I don't think it is needed anymore? (Additionally I decided to always include the seen messages up to the fail in the test failure, so maybe that also helps)

Comment (ktoso) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/TimerSpec.scala]: I can't include any hints for the cases: timeout, match error, only for explicit fail

Comment (johanandren) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/TimerSpec.scala]: Fair point, I'll add it

File: akka-testkit-typed/src/main/scala/akka/testkit/typed/TestKit.scala [hunk 4]
```diff
@@ -17,6 +17,14 @@ import scala.util.control.NoStackTrace
  */
 case class TE(message: String) extends RuntimeException(message) with NoStackTrace
 
+/**
+ * Not for user extension.
+ *
+ * Instances are available from `FishingOutcomes` in the respective dsls: [[akka.testkit.typed.scaladsl.FishingOutcomes]]
```
Comment (chbatey) [inline akka-testkit-typed/src/main/scala/akka/testkit/typed/TestKit.scala]: :+1:  to separate APIs. That is what we said we'd do from now on wasn't it? Separate interfaces with one impl. Would be nice to remove the finite durations from the java api as well.

**Review by chbatey — APPROVED** [review]
LGTM, up to you if you want to separate out the APIs, or we can create a ticket?

File: akka-testkit-typed/src/main/scala/akka/testkit/typed/TestKit.scala (hunk 4 above)
Comment (johanandren) [reply akka-testkit-typed/src/main/scala/akka/testkit/typed/TestKit.scala]: Created #24542 to track separating the APIs

**Pull request merged**